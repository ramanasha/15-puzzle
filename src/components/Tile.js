import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';

class Tile extends Component {
  static propTypes = {
    tileId: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    left: PropTypes.number.isRequired,
    top: PropTypes.number.isRequired,
    number: PropTypes.number,
    onClick: PropTypes.func,
    correct: PropTypes.bool,
    visible: PropTypes.bool,
  };

  static defaultProps = {
    number: 0,
    correct: false,
    visible: true,
  };

  render() {
    const { className, number, visible } = this.props;
    const classes = classnames({
      [className]: true,
      'tile-invisible': !visible,
    });

    return (
      <div className={classes} onClick={this.onClick}>
        <span className="tile-number">{number}</span>
      </div>
    );
  }

  onClick = () => {
    this.props.onClick({
      number: this.props.number,
    });
  };
}

export default styled(Tile)`
  border: 1px solid #FFD1AA;
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  left: ${props => props.left}px;
  top: ${props => props.top}px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: ${props => props.correct ? '#226666' : '#D4726A'};
  transition-property: top, left, background-color;
  transition-duration: .300s;
  transition-timing-function: ease-in;

  .tile-number {
    color: #FFD1AA;
    font-weight: 400;
    fontSize: 1.8em;
    user-select: none;
  }

  &.tile-invisible {
    display: none;
  }
`;

// Export this for easier unit-testing
export { Tile as Component };

export const propTypes = Tile.propTypes;
