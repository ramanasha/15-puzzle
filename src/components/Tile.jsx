import React, { Component } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class Tile extends Component {
  onClick = () => {
    this.props.onClick({
      number: this.props.number
    });
  };

  render() {
    const { className, number, visible } = this.props;
    const classes = classnames({
      [className]: true,
      'tile-invisible': !visible
    });

    return (
      <div className={classes} onClick={this.onClick}>
        <span className="tile-number">{number}</span>
      </div>
    );
  }
}

Tile.propTypes = {
  tileId: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  left: PropTypes.number.isRequired,
  top: PropTypes.number.isRequired,
  number: PropTypes.number,
  onClick: PropTypes.func.isRequired,
  correct: PropTypes.bool,
  visible: PropTypes.bool
};

Tile.defaultProps = {
  number: 0,
  correct: false,
  visible: true
};

export default styled(Tile)`
  border: 1px solid #ffd1aa;
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  left: ${props => props.left}px;
  top: ${props => props.top}px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: ${props => (props.correct ? '#226666' : '#D4726A')};
  transition-property: top, left, background-color;
  transition-duration: 0.3s;
  transition-timing-function: ease-in;

  .tile-number {
    color: #ffd1aa;
    font-weight: 400;
    font-size: 1.8em;
    user-select: none;
  }

  &.tile-invisible {
    display: none;
  }
`;

// Export this for easier unit-testing
export { Tile as Component };

export const propTypes = Tile.propTypes;
