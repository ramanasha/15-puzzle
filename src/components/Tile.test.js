import renderer from 'react-test-renderer';
import React from 'react';
import Tile from './Tile';
import { Component as ReactTile } from './Tile';
import { shallow } from 'enzyme';

describe('Tile', () => {
  describe('render tests', () => {
    it('should have a computed class prop', () => {
      const wrapper = shallow(
        <Tile
          tileId={50}
          left={0}
          top={0}
          number={13}
          width={90}
          height={90}
        />,
      );
      expect(wrapper.props().className).not.toBeUndefined();
    });

    it('should render correctly given custom number', () => {
      const tree = renderer
        .create(
          <Tile
            tileId={50}
            left={0}
            top={0}
            number={13}
            width={90}
            height={90}
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render correctly given default number', () => {
      const tree = renderer
        .create(
          <Tile tileId={18} left={-40} top={-20} width={120} height={120} />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render correctly given correct prop', () => {
      const tree = renderer
        .create(
          <Tile
            tileId={13}
            left={0}
            top={50}
            correct={true}
            width={90}
            height={90}
            number={23}
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render correctly given visible prop', () => {
      const tree = renderer
        .create(
          <Tile
            tileId={4}
            left={30}
            top={0}
            visible={false}
            width={30}
            height={30}
            number={1}
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('callback tests', () => {
    it('should call onClick function passed', () => {
      const onClick = jest.fn();
      const number = 42;

      const wrapper = shallow(
        <ReactTile
          tileId={4}
          left={30}
          top={0}
          correct={true}
          width={30}
          height={30}
          number={number}
          onClick={onClick}
        />,
      );

      wrapper.instance().onClick();
      expect(onClick).toBeCalled();
    });
  });
});
