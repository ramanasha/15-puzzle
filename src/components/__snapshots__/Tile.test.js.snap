// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile render tests should render correctly given correct prop 1`] = `
<div
  className="sc-bdVaJa dFyZnU"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    23
  </span>
</div>
`;

exports[`Tile render tests should render correctly given custom number 1`] = `
<div
  className="sc-bdVaJa jodVcS"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    13
  </span>
</div>
`;

exports[`Tile render tests should render correctly given default number 1`] = `
<div
  className="sc-bdVaJa hYSZle"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    0
  </span>
</div>
`;

exports[`Tile render tests should render correctly given visible prop 1`] = `
<div
  className="sc-bdVaJa clEXFS tile-invisible"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    1
  </span>
</div>
`;
